tonic/src/transport/server/incoming.rs

```diff
@@ -27,7 +27,19 @@ where
         let mut incoming = pin!(incoming);
 
         while let Some(item) = incoming.next().await {
-            yield item.map(ServerIo::new_io)?
+            if let Err(e) = item {
+                if let Some(e) = Into::<crate::Error>::into(e).downcast_ref::<std::io::Error>() {
```

Comment:
I think this is broadly acceptable but I have bunch of stylistic suggestions.

First, use a match for the top-level error extraction:

```rust
let err = match item {
    Ok(io) => yield Ok(ServerIo::new_io(io)), // Or however this is spelled
    Err(err) => err,
};

..
```

This reduces rightward drift.

Second, I would expect we can just call `err.into()` and don't need the whole `Into::<crate::Error>::into(e)` dance?

Third, for the debug messages, I think we should not use an explicit `message =` name but use the general form which is having the message after the "structured" argument (so `tracing::debug!(error = %e, "accept loop error");`). Note that error messages should be all-lowercase and not have any `.` at the end.

Fourth, please import `io` at the top of the file so we can drop the `std::` prefixes here.

Reply:
> I think this is broadly acceptable but I have bunch of stylistic suggestions.
> 
> First, use a match for the top-level error extraction:
> 
> ```rust
> let err = match item {
>     Ok(io) => yield Ok(ServerIo::new_io(io)), // Or however this is spelled
>     Err(err) => err,
> };
> 
> ..
> ```
> 
> This reduces rightward drift.
> 
> Second, I would expect we can just call `err.into()` and don't need the whole `Into::<crate::Error>::into(e)` dance?
> 
> Third, for the debug messages, I think we should not use an explicit `message =` name but use the general form which is having the message after the "structured" argument (so `tracing::debug!(error = %e, "accept loop error");`). Note that error messages should be all-lowercase and not have any `.` at the end.
> 
> Fourth, please import `io` at the top of the file so we can drop the `std::` prefixes here.

Personally I don't have much of an opinion on these style issues, so I just changed things as requested.

There is one thing I really didn't like though, which was error extraction, which would have looked like this:

```
let err = match item {
    Ok(io) => { yield Ok(ServerIo::new_io(io)); continue },
    Err(err) => err,
};

do more stuff with err, then break;
```

So instead, I went with clearly defined Ok and Err branches (horizontal drift is IMHO acceptable). I'm quite happy with the readability of the outcome (especially having only one continue statement, pointing out the exception, and no break statements). This should also be relatively easy to parse to people who are not familiar with all language features + the diff to the master branch is very easy to understand.